modules/base/src/algorithm/volume/volumevoronoi.cpp

```diff
@@ -29,69 +29,180 @@
 
 #include <modules/base/algorithm/volume/volumevoronoi.h>
 
+#include <array>
+
 namespace inviwo {
 namespace util {
 
-std::shared_ptr<Volume> voronoiSegmentation(
+namespace detail {
+template <Wrapping X, Wrapping Y, Wrapping Z, typename T>
+auto distance2(const glm::vec<3, T>& a, const glm::vec<3, T>& b,
+               [[maybe_unused]] const glm::vec<3, T>& size) {
+    auto delta = b - a;
+
+    if constexpr (X == Wrapping::Repeat) {
+        if (delta.x > 0.5 * size.x) delta.x -= size.x;
+        if (delta.x < -0.5 * size.x) delta.x += size.x;
+    }
+
+    if constexpr (Y == Wrapping::Repeat) {
+        if (delta.y > 0.5 * size.y) delta.y -= size.y;
+        if (delta.y < -0.5 * size.y) delta.y += size.y;
+    }
+
+    if constexpr (Z == Wrapping::Repeat) {
+        if (delta.z > 0.5 * size.z) delta.z -= size.z;
+        if (delta.z < -0.5 * size.z) delta.z += size.z;
+    }
+
+    return glm::length2(delta);
+}
+
+template <typename Index, typename Functor, Index... Is>
+constexpr auto build_array_impl(Functor&& func, std::integer_sequence<Index, Is...>) noexcept {
+    return std::array{func(std::integral_constant<Index, Is>{})...};
+}
+
+template <std::size_t N, typename Index = size_t, typename Functor>
+constexpr auto build_array(Functor&& func) noexcept {
+    return build_array_impl<Index>(std::forward<Functor>(func),
+                                   std::make_integer_sequence<Index, N>());
+}
+
+}  // namespace detail
+
+template <Wrapping X, Wrapping Y, Wrapping Z>
+void voronoiSegmentationImpl(const size3_t volumeDimensions, const mat4& indexToModelMatrix,
```

Comment:
I guess there wasn't any good way to merge `voronoiSegmentationImpl` and `weightedVoronoiSegmentationImpl`..?